pkg/controller/azvolume.go

```diff
@@ -65,6 +69,49 @@ type reconcileAzVolume struct {
 var _ reconcile.Reconciler = &reconcileAzVolume{}
 
 func (r *reconcileAzVolume) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	result, err := r.HandlePV(ctx, request)
+	if !result.Requeue && errors.IsNotFound(err) {
+		return r.HandleAzVolume(ctx, request)
+	}
+	return result, err
+}
+
+func (r *reconcileAzVolume) HandlePV(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	var pv corev1.PersistentVolume
+	if err := r.client.Get(ctx, request.NamespacedName, &pv); err != nil {
+		if errors.IsNotFound(err) {
+			return reconcile.Result{}, err
+		}
+		klog.Errorf("failed to get PV (%s): %v", request.Name, err)
+		return reconcile.Result{Requeue: true}, err
+	}
+
+	if pv.Spec.CSI == nil || pv.Spec.CSI.Driver != DriverName {
+		return reconcile.Result{}, nil
+	}
+
+	var azVolume v1alpha1.AzVolume
+	if err := r.client.Get(ctx, types.NamespacedName{Namespace: r.namespace, Name: pv.Name}, &azVolume); err != nil {
+		// if underlying PV was found but AzVolume was not. Might be due to stale cache
+		klog.V(5).Infof("failed to get AzVolume (%s): %v", pv.Name, err)
+		return reconcile.Result{}, nil
+	}
+
+	if azVolume.Status != nil {
+		if pv.Status.Phase == corev1.VolumeReleased && azVolume.Status.Phase == v1alpha1.VolumeBound {
+			updated := azVolume.DeepCopy()
+			updated.Status.Phase = v1alpha1.VolumeReleased
+
+			if err := r.client.Update(ctx, updated, &client.UpdateOptions{}); err != nil {
```

Comment:
Shouldn't we cleanup the AzVolumeAttachment after doing this? Looks like HandleAzVolume or CleanUpAzVolumeAttachment will not be called after this.

Reply:
This updates the AzVolume object, and this update event will be added to the workqueue. Then, this will eventually be handled in the next reconciliation as follows:
https://github.com/abhisheksinghbaghel/azuredisk-csi-driver/blob/8e8f802c4209d5629e7837c1d7d055a43398065a/pkg/controller/azvolume.go#L143

Reply:
Would it be better to cleanup the attachments instead of waiting for the next cycle?

Reply:
Once it enters the `Reconcile` function, we no longer have access to the PV object's previous state. So the reconciler doesn't know if the update event was triggered by the transition of PV's phase from `VolumeBound` to `VolumeReleased` or something else. On that note, if some other field of PV was updated while the phase is `VolumeReleased`, the reconciler may end up making unncessary `CleanUpAzVolumeAttachments` calls.